utils/global-state-update-gen/src/generic.rs

```diff
@@ -0,0 +1,284 @@
+    let config_path = matches.value_of("config_file").unwrap();
+
+    let config_bytes = fs::read(config_path).expect("couldn't read the config file");
+    let config: Config = toml::from_slice(&config_bytes).expect("couldn't parse the config file");
+
+    let builder = LmdbWasmTestBuilder::open_raw(data_dir, Default::default(), state_hash);
+
+    update_from_config(builder, config);
+}
+
+pub(crate) fn get_update<T: StateReader>(reader: T, config: Config) -> BTreeMap<Key, StoredValue> {
+    let mut state_tracker = StateTracker::new(reader);
+
+    process_transfers(&mut state_tracker, &config.transfers);
+
+    update_account_balances(&mut state_tracker, &config.accounts);
+
+    update_auction_state(
+        &mut state_tracker,
+        &config.accounts,
+        config.only_listed_validators,
+    );
+
+    state_tracker.get_entries()
+}
+
+pub(crate) fn update_from_config<T: StateReader>(reader: T, config: Config) {
+    let update = get_update(reader, config);
+    for (key, value) in update {
+        print_entry(&key, &value);
+    }
+}
+
+fn process_transfers<T: StateReader>(state: &mut StateTracker<T>, transfers: &[Transfer]) {
+    for transfer in transfers {
+        state.execute_transfer(transfer);
+    }
+}
+
+fn update_account_balances<T: StateReader>(
+    state: &mut StateTracker<T>,
+    accounts: &[AccountConfig],
+) {
+    for account in accounts {
+        let target_balance = if let Some(balance) = account.balance {
+            balance
+        } else {
+            continue;
+        };
+        let account_hash = account.public_key.to_account_hash();
+        if let Some(account) = state.get_account(&account_hash) {
+            state.set_purse_balance(account.main_purse(), target_balance);
+        } else {
+            state.create_account(account_hash, target_balance);
+        }
+    }
+}
+
+fn update_auction_state<T: StateReader>(
+    state: &mut StateTracker<T>,
+    accounts: &[AccountConfig],
+    only_listed_validators: bool,
+) {
+    // Read the old SeigniorageRecipientsSnapshot
+    let (validators_key, old_snapshot) = state.read_snapshot();
+
+    // Create a new snapshot based on the old one and the supplied validators.
+    let new_snapshot = if only_listed_validators {
+        gen_snapshot_only_listed(
+            *old_snapshot.keys().next().unwrap(),
+            old_snapshot.len() as u64,
+            accounts,
+        )
+    } else {
+        gen_snapshot_from_old(old_snapshot.clone(), accounts)
+    };
+
+    if new_snapshot != old_snapshot {
+        // Save the write to the snapshot key.
+        state.write_entry(
+            validators_key,
+            StoredValue::from(CLValue::from_t(new_snapshot.clone()).unwrap()),
+        );
+
+        let validators_diff = validators_diff(&old_snapshot, &new_snapshot);
+
+        add_and_remove_bids(
+            state,
+            &validators_diff,
+            &new_snapshot,
+            only_listed_validators,
+        );
```

Comment:
Just to make sure I understand correctly: This will remove all relevant bids, because anyone who made a high enough bid will be present in `old_snapshot`, because the last block before the restart is always a switch block?

Reply:
It doesn't look at the old snapshot, though - it looks at _all_ the bids that are in the global state, including those that weren't winning auctions, and so their public keys wouldn't be in the snapshot.